Add CosmWasm specification (#2539)
ComposableFi/composable: specifications/cosmwasm/SPEC.md

```
Composable Finance
Hussein Ait Lahcen, Abdullah Eryuzlu, Karel L. Kubat, Antonio M. Larriba
2022-11-04
```

# Abstract
Smart contracts provide the capability for third parties to run deterministic, reproducible scripts on blockchains, triggered by extrinsics events such as transactions.
The most prevalent virtual machine for executing contracts is the Ethereum Virtual Machine (EVM), but many others exist.
Here we specify the inner workings of our CosmWasm Virtual Machine, an actor-based VM for web assembly contracts.

# Status of This Memo

> This specification describes an already implemented virtual machine. New features are not accepted in this document but require a separate RFC.

Author: The idea was to add it under the specifications folder ([specifications](https://app.clickup.com/t/20465559/XCVM-16).
Should we move it @KaiserKarel ? Or simply reference the spec from the internal documentation?

Instantiation is the process of creating a new contract singleton from the uploaded code and instantiation parameters, deriving a unique identifier to call the contract (the address).
Migrations provide a way to upgrade the smart-contract code of an already instantiated contract.
This functionality is opt-in and can be disabled by contract authors.

As of the moment of writing this document, state rent is not part of the CosmWasm standard or our implementation.

## 1.1. Document Structure
Section 1 provides an overview of the specification and its notation.
The rest of the manuscript, is structured as follows.

Section 2 introduces the concept of smart contracts, the interfaces and functionalities they are REQUIRED to support in order to be compliant.
Section 3 defines the host functions supported by our specification.
In Section 4, we cover the virtualization details of the implementation, and, Section 5 briefly defines the serialization requirements.
Section 6 depicts the IBC compatibility mechanisms of our implementation.
In Section 7, we provide specific details about the pallet implementation of our specification.

> Section 8 covers security considerations that every developer MUST follow when using our specification.

Reviewer: Like this? https://github.com/ComposableFi/composable/pull/2539/commits/b1ab5444666f556fbe52b3af1d1ecbb4e83ceddd

> Section 8 covers security considerations that every developer MUST follow when using our specification.

Reviewer 2: exactly

The contract is not required to wipe previously allocated memory.
It is only marked, by the virtual machine, as no longer in use.
The `deallocate` export MUST be present in the contract binary.

```rust
extern "C" fn deallocate(pointer: u32);
```

Compiling contracts with alternative allocators and never deallocating can yield significant performance improvements and reductions in contract size.

## 2.5. Instantiation
After uploading a contract, the virtual machine will instantiate it by passing the instantiation message.
Instantiate export MUST be present in the contract binary.

```rust
extern "C" fn instantiate(env_ptr: u32, info_ptr: u32, msg_ptr: u32) -> u32;
```

The virtual machine MUST call `instantiate` before allowing `execute` and `query` access.
Implementors and contract authors MAY choose to lazily execute `instantiate` upon the first `execute` or `query` call.
Our virtual machine eagerly executes it.

The environment pointer (`env_ptr`) contains the current state information.
The message information pointer (info_ptr) contains the sender (public key) along with the tokens that have been sent for the contract call.

> See sections 2.4.1 and 2.4.2 for the exact definitions of these pointers.

Author: Yes

Instead, a double indirection exists, where the pointer points to a `Region`.
Regions are readable and writable sections of wasm memory.

```rust
#[repr(C)]
pub struct Region {
    pub offset: u32,
    pub capacity: u32,
    pub length: u32,
}
```

`Regions` are decoded using their C representation, not by JSON deserialization.
They are safe to return over FFI boundaries.

A `Region`'s `offset` is the number of bytes from the start of the wasm linear memory.
The `capacity` is the maximum number of bytes that the virtual machine is allowed to write to that region, used for operations that write to memory.
The `length` is the number of bytes that MAY be read from the `offset`, prepared by the smart contract.
Implementors MUST never read past `length` bytes.

## 2.6. Execute

After instantiation, the contract MAY receive `execute` calls.
The `execute` export MUST be present in the contract binary.

```rust
extern "C" fn execute(env_ptr: u32, info_ptr: u32, msg_ptr: u32) -> u32;
```

The virtual machine MAY use `execute` when a transaction calls a contract, or when the contract is called by another contract or module.

> All changes to storage MUST be actuated by the virtual machine during the `execute` call. See sections 2.4.1 and 2.4.2 for the exact definitions of the `env_ptr` and `info_ptr` pointers. The `msg_ptr` is defined by the contract itself and by convention, JSON encoded.

Reviewer: Done for all cases following this one. The special cases 2.1.1., 2.11., and 2.1.2., 2.12. are solved by linking directly in the text.

Contracts MUST enable the **ibc3** feature to have access to the `relayer` field of `IbcPacketTimeoutMsg`.

Contracts MAY return an error or an [IbcBasicResponse](https://github.com/CosmWasm/cosmwasm/blob/6033d91aab8dae99c0fe3b4ecf649242b21fde8a/packages/std/src/ibc.rs#L486) from the call.

```rust
pub struct IbcBasicResponse<T = Empty> {
    pub messages: Vec<SubMsg<T>>,
    pub attributes: Vec<Attribute>,
    pub events: Vec<Event>,
}
```

Where `messages` are the sub-messages to process.

Where `attributes` are the attributes associated with the default event that will be yield for the call.

Where `events` are custom events that can be yield along the default event.

# 3. Host Functions
The virtual machine MUST provide the host functions listed below.
A host function is provided by the virtual machine and callable by contracts to interact with the blockchain itself, very much like syscalls in operating system contexts.
The next sections define each function in more depth.

> The `readonly` column determine whether the operation is allowed to alter the storage or not -- a readonly operation altering the storage result in a readonly violation.

Reviewer: @Fantoni0 please confirm this is not intended to represent some syntax/semantics.

> The `readonly` column determine whether the operation is allowed to alter the storage or not -- a readonly operation altering the storage result in a readonly violation.

Author: It was not intended, just an spelling error.

```rust
pub enum ReplyOn {
    /// Always perform a callback after SubMsg is processed
    Always,
    /// Only callback if SubMsg returned an error, no callback on success case
    Error,
    /// Only callback if SubMsg was successful, no callback on error case
    Success,
    /// Never make a callback - this is like the original CosmosMsg semantics
    Never,
}
```
If the caller decides to not handle the error, its state will be reverted. 

### 4.2.2 Query
Queries are read-only messages.
They are used to query the state of another actor in the system.

```rust
#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
#[serde(rename_all = "snake_case")]
pub enum QueryMsg {
  // ResolveAddress returns the current address that the name resolves to
  ResolveRecord { name: String },
  Config {},
} 
```

## 4.3. Calling a Contract
Contract to contract calls are structured through messaging.
Calling another contract requires messages to go through a `Dispatcher` that maintains the state related to the calls.
The purpose of the dispatcher is twofold: first, it prevents contracts from calling other contracts directly to prevent reentrancy attacks.

> Second , it helps to maintain atomicity and revert if a call to another contract fails.

Author: It could be added. However, we tried to limit the list of terms and definitions to Cosmwasm only terms.
We assumed that general blockchain terms would be already known by the potential reader. Unless atomicity has a special connotation in CosmWasm, I would not add it.

## 4.4. Contract Call Graph
A call graph is a flow graph that represents the calling relationships between actors and theirs subroutines. 
For CosmWasm this means the call graph represents the transactional flow of messages between smart contracts.
These dependencies can grow fast, specially in cross-chain scenarios. 
Precedence is determined using a width-first approach.

> This means if an actor needs to process two messages `M1` and `M2` from actor `A1`, and two messages  `N1` and `N2` from actor `A2, the order will be: `M1`, `N1`, `M2`, `N2`. 

Reviewer: fixed here https://github.com/ComposableFi/composable/pull/2539/commits/494c77956d80861b360dff36f5b9fae03ac350e3

### 6.2.3. IBC Channel Close

Upon reception of the `IbcMsg::CloseChannel` variant submessage by the VM, the according `ibc_channel_close` host function is called.

```rust
#[cfg(feature = "stargate")]
fn ibc_close_channel(&mut self, channel_id: String) -> Result<(), Self::Error>;
```

Implementors MUST verify that the `channel_id` exists and that it has previously been opened (see section 2.11.).

Implementors MAY forward the call to a specific module in charge of handling IBC operations.

# 7. Pallet-CosmWasm
In this section we provide actual implementation details of the specification as a Substrate pallet.
We cover some of the theoretical topics that were presented before, and provide real code that exemplifies how they can be developed in a pallet.

## 7.1. Gas Metering
When a contract is uploaded through the `upload` extrinsic, its code is instrumented.
This instrumentation process adds both gas metering and stack height limit to the uploaded code.

Proper gas metering is achieved in two steps:
1. **Benchmarking**: Each instruction is benchmarked to determine the cost of its execution.

> 2. **Injecting Gas Metering**: Gas metering is injected to the code by using [wasm_instrument](https://github.com/paritytech/wasm-instrument). This process injects a call to the gas function for every execution or path/code block (function call, if, else, etc.) with the associated execution cost parameter. Therefore, it computes the overall total cost and makes sure that every code block is paid before getting executed.

Reviewer: ```suggestion
2. **Injecting Gas Metering**: Gas metering is injected to the code by using [wasm_instrument](https://github.com/paritytech/wasm-instrument). This process injects a call into the gas function for every execution or path/code block (function call, if, else, etc.) with the associated execution cost parameter. Therefore, it computes the total cost and ensures that every code block is paid before getting executed.
```
Nit

> Then, everytime a contract entrypoint is called, the pallet checks if the instrumentation version is up-to-date.

Reviewer: Fixed here with "whenever"
https://github.com/ComposableFi/composable/pull/2539/commits/a1b6ffebb6cb4ea104be7e43176ec9edc43ec192

In case not, the code gets re-instrumented to ensure proper gas metering.

## 7.2. Uploading Contracts
The `upload` extrinsic is used to upload smart contracts´ `code` to the pallet.

### 7.2.1. Definition
```rust
#[pallet::weight(T::WeightInfo::upload(code.len() as u32))]
pub fn upload(origin: OriginFor<T>, code: ContractCodeOf<T>) -> DispatchResultWithPostInfo;
```

### 7.2.2. Execution Flow
1. Check if the `code` is already uploaded.
2. Reserve `length(code) * deposit_per_byte` amount of native asset.
3. Check if the wasm code is valid.
4. Do the instrumentation by injecting gas metering and stack height limit.
5. Assign a code id to the `code`. This `code id ` is incremented on each upload.
6. Deposit the upload event.

```rust
pub enum Event<T: Config> {
	Uploaded {
		code_hash: CodeHashOf<T>,
		code_id: CosmwasmCodeId,
	}
```

### 7.2.3. Fee
Fees depend linearly on the size of the code.

## 7.3. Instantiating a Contract

> THe `instantiate` extrinsic is used to instantiate a smart contract.

Reviewer: ```suggestion
The `instantiate` extrinsic is used to instantiate a smart contract.
```
Nit

### 7.3.1 Definition
```rust
#[pallet::weight(T::WeightInfo::instantiate(funds.len() as u32).saturating_add(*gas))]
pub fn instantiate(
	origin: OriginFor<T>,
	code_identifier: CodeIdentifier<T>,
	salt: ContractSaltOf<T>,
	admin: Option<AccountIdOf<T>>,
	label: ContractLabelOf<T>,
	funds: FundsOf<T>,
	gas: u64,
	message: ContractMessageOf<T>,
) -> DispatchResultWithPostInfo;
```

`origin` will be the `sender` from the contracts' perspective.

Our goal is to create a deterministic smart contract environment.
Hence, we are not only using `code id` to identify a code.
Since `code id` depends on the current state of the chain, users won't be able to deterministically identify their code.
So we created `CodeIdentifier` which makes users able to also identify their code by using `CodeHash`.
And the corresponding `code id` is fetched internally.

> This feature comes handy when users want to batch their `upload + instantiate` calls, or they do any kind of scripting to upload and run the contracts.

Reviewer: ```suggestion
This feature comes in handy when users want to batch their `upload + instantiate` calls, or they do any kind of scripting to upload and run the contracts.
```

The total fee can be computed as follows.
```
base_instantiate_fee + (fee_per_fund * length(funds)) + executed_instruction_costs
```
The remaining gas is refunded after execution.

# 8. Security considerations
There have been a number of vulnerabilities in different smart-contract execution environments, often enabled because of small implementation mistakes. Each of these has unique circumstances.
Instead of going into each of these in-depth, we recommend:

1. Implementors to always use third-party auditors to verify their implementations.

2. Make use of fuzzing and verification frameworks to proof correctness. Some libraries we recommend are:

    - [kani](https://github.com/model-checking/kani)
    - [proptest](https://docs.rs/proptest/latest/proptest)
    - [fuzzing](https://github.com/rust-fuzz/cargo-fuzz)

## 8.1. Previous Vulnerabilities

- https://halborn.com/halborn-discovers-zero-day-vulnerability-in-cosmwasm

# 9. Contributors
Contributors that made our CosmWasm implementation possible are here listed in alphabetical order:

- Hussein Ait Lahcen - XVCM Principal Engineering - Composable Finance
- Abdullah Eryuzlu - Rust Developer - Composable Finance

> - Cor Pruijs - Rust Developer - Composable Finance

Reviewer: @KaiserKarel @Fantoni0 Should this be the same as the mentions in the doc header? 
https://github.com/ComposableFi/composable/pull/2539/files#diff-2a2ac51d9838b658d158b1bc767b2a769a219a9721a251e8dbfbfacd850ebaf2R3

> - Cor Pruijs - Rust Developer - Composable Finance

Author: I think the doc header contains the people that contributed to the specification writing. While the Contributors section, include people that made the VM possible.
